### ISSUE TITLE
Cling embedding demo is broken on Windows VS2015-2017

### ISSUE THREAD

**Issue by danhambleton** [root-project/cling#200]
After successfully building LLVM/Clang/Cling and playing around with the Cling driver, I tried out the embedding demo. It compiles successfully, and the Interpreter does get created, but a bunch of cling symbols are not found:

```
IncrementalExecutor::executeFunction: symbol '?setValueNoAlloc@internal@runtime@cling@@YAXPEAX00D_K@Z' unresolved while linking [cling interface function]!
You are probably missing the definition of void __cdecl cling::runtime::internal::setValueNoAlloc(void * __ptr64,void * __ptr64,void * __ptr64,char,unsigned __int64)
Maybe you need to load the corresponding shared library?
aGlobal is 1633814599152
```

I have been trying to figure this out for a couple days now, so any help would be most appreciated! 

More errors here:
![image](https://user-images.githubusercontent.com/1939589/33901621-6043018e-df40-11e7-86b6-17f1a6fc72c4.png)

**Comment by bellenot** [comment]
Can you try to add the missing symbols (the mangled ones) in the `cling_exports` list in tools/driver/CMakeLists.txt, and rebuild, to see if it solves the issue?

**Comment by danhambleton** [comment]
So, adding the missing symbols from the driver did not solve the issue. However, I'm sure it is required step. However, manually adding in the symbols that the demo was complaining about, DOES seem to fix it. 
![image](https://user-images.githubusercontent.com/1939589/33945134-2329d49c-dfec-11e7-8628-31bb8c78eedc.png)

I added these symbols: 

```
?setValueNoAlloc@internal@runtime@cling@@YAXPEAX00D_K@Z
?setValueNoAlloc@internal@runtime@cling@@YAXPEAX00DM@Z
?getAnotherGlobal@@YAMXZ
?setAnotherGlobal@@YAXM@Z
```

So that is good. My concern now is that this just treats the current missing symbols. How can I be sure that I have got them all?

Anyway, it is progress!

**Comment by danhambleton** [comment]
Ok, I am confused again: I thought those symbols were cling utility functions, when in fact they are defined in the demo (at least the "getAnotherGlobal"):

![image](https://user-images.githubusercontent.com/1939589/33945659-ac602814-dfed-11e7-9b54-5023b12e2ac5.png)

I guess the demo assumes a "no mangling" platform? If I wrapped those functions in "extern C", would they link properly?

**Comment by danhambleton** [comment]
Sorry for the barrage of comments - last one for today I promise ;)

I tested out some other code that DOES WORK in the cling driver, but fails in the demo:

![image](https://user-images.githubusercontent.com/1939589/33946024-aadbb9da-dfee-11e7-898b-612f1b8629c8.png)

**Comment by Axel-Naumann** [comment]
Right. The problem is that the cling-demo has these symbols as part of the executable. We need to convince the executable to be linked "as if it were a shared library" - we do the same for cling.

I.e. the mechanism we use to export symbols from cling.exe needs to be used also for cling-demo.exe. Bertrand, can you help? 🆘

**Comment by bellenot** [comment]
Oh, sorry, I overlooked at the issue... Where is the cling-demo.exe' CMakeLists.txt?

**Comment by danhambleton** [comment]
@bellenot Yes, this does work - thanks! See above for more comments, though.

**Comment by bellenot** [comment]
@danhambleton Cool, thanks for the feedback! And can you summarize what are the remaining issues (or concerns) you still have?

**Comment by danhambleton** [comment]
Well, from my test/usage:
- While some symbols have been correctly linked in, others are still missing from the demo (for instance: `cling_runtime_internal_throwIfInvalidPointer`. How can we get them all?
- In general, the demo of embedding cling and the cling driver behave very differently. How can we get feature parity? The cling driver is far more stable and can handle all kinds of code, whereas the demo can't handle some pretty basic constructs ([like this](https://user-images.githubusercontent.com/1939589/33946024-aadbb9da-dfee-11e7-898b-612f1b8629c8.png))

**Comment by bellenot** [comment]
So in principle, all the symbols should be available, but I'll have to check why some are still missing. For the time being, adding them in the export_list (as in the code I posted) should solve the issue. So can you also add `cling_runtime_internal_throwIfInvalidPointer` in the `tcling_demo_exports` list and tell me if it works?
For the second point, keep in mind that cling is is not fully functional on Windows. And for the error your link points to, it is a known issue.

**Comment by danhambleton** [comment]
I actually find that Cling works pretty well on Windows! The thing that is really holding me back is that when I embed Cling I can't run code like in the link, but that exact same code DOES run in the Cling driver. Any tips on why that might be? 
I will try adding that symbol. It doesn't seem to have a mangled name, so not sure...
In any case, thanks so much for all the help!

**Comment by bellenot** [comment]
Adding the symbol should work (it works for the ROOT project), and I have no idea why those symbols are not visible when embedding cling... For example, in ROOT we export those symbols:
```
   set(cling_exports ??_7type_info@@6B@
      cling_runtime_internal_throwIfInvalidPointer
      ?_Facet_Register@std@@YAXPAV_Facet_base@1@@Z
      ??1Value@cling@@QAE@XZ
      ?isValid@Value@cling@@QBE_NXZ
      ?AssertOnUnsupportedTypeCast@Value@cling@@IBEXXZ
      ?setValueNoAlloc@internal@runtime@cling@@YAXPAX00D@Z
      ?setValueNoAlloc@internal@runtime@cling@@YAXPAX00DM@Z
      ?setValueNoAlloc@internal@runtime@cling@@YAXPAX00DN@Z
      ?setValueNoAlloc@internal@runtime@cling@@YAXPAX00DO@Z
      ?setValueNoAlloc@internal@runtime@cling@@YAXPAX00D_K@Z
      ?setValueNoAlloc@internal@runtime@cling@@YAXPAX00DPBX@Z
      ?setValueWithAlloc@internal@runtime@cling@@YAPAXPAX00D@Z
      ?printValue@cling@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PB_N@Z
      ?printValue@cling@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PAPBX@Z
      ??$toUTF8@D@cling@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QBDID@Z
      ?EvaluateDynamicExpression@internal@runtime@cling@@YA?AVValue@3@PAVInterpreter@3@PAVDynamicExprInfo@123@PAVDeclContext@clang@@@Z
  )
```
(but ROOT is built in 32 bit mode)

**Comment by danhambleton** [comment]
Hey, that worked!! Here is the CMake file for the embedding demo that is now working for me:

```
#------------------------------------------------------------------------------
# CLING - the C++ LLVM-based InterpreterG :)
#
# This file is dual-licensed: you can choose to license it under the University
# of Illinois Open Source License or the GNU Lesser General Public License. See
# LICENSE.TXT for details.
#------------------------------------------------------------------------------

cmake_minimum_required(VERSION 3.5)

# Keep symbols for JIT resolution
set(LLVM_NO_DEAD_STRIP 1)

# Cling needs at least C++11; so does this demo.
set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

if (CMAKE_PROJECT_NAME)
  # Building as part of cling; all CMake variables are set.
else()
  # Building as separate project.
  project(cling-demo)

  # This project needs cling.
  find_package(cling REQUIRED)
endif()

# The project has one binary:
add_executable(cling-demo cling-demo.cpp)

if(MSVC)
  set_target_properties(cling-demo PROPERTIES WINDOWS_EXPORT_ALL_SYMBOLS 1)

  # Internal string
  set(cling_exports ?kEmptyCollection@valuePrinterInternal@cling@@3QEBDEB)

  # RTTI/C++ symbols
  set(cling_exports ${cling_exports} ??_7type_info@@6B@
    ?__type_info_root_node@@3U__type_info_node@@A
    ?nothrow@std@@3Unothrow_t@1@B
    ?_Facet_Register@std@@YAXPEAV_Facet_base@1@@Z
  )

  # Compiler added symbols for static variables. NOT for VStudio < 2015
  set(cling_exports ${cling_exports} _Init_thread_abort _Init_thread_epoch
    _Init_thread_footer _Init_thread_header _tls_index
  )

  # new/delete variants needed when linking to static msvc runtime (esp. Debug)
  set(cling_exports ${cling_exports}
    #??2@YAPEAX_KPEBDH@Z  ## not used in cling
    #??_U@YAPEAX_KPEBDH@Z ## not used in cling
    ??2@YAPEAX_K@Z
    ??3@YAXPEAX@Z
    ??_U@YAPEAX_K@Z
    ??_V@YAXPEAX@Z
  )

  # Most (if not all) of these MSVC decided are inlines that aren't exported
  # TODO: There are probably some missing. Why?
  set(cling_exports ${cling_exports} ?print@Decl@clang@@QEBAXAEAVraw_ostream@llvm@@I_N@Z
    ?getAsString@QualType@clang@@SA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PEBVType@2@VQualifiers@2@@Z
    ??6raw_ostream@llvm@@QEAAAEAV01@PEBX@Z
    ?getQualifiedNameAsString@NamedDecl@clang@@QEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ
    ?insert_imp_big@SmallPtrSetImplBase@llvm@@AEAA?AU?$pair@PEBQEBX_N@std@@PEBX@Z
    ?getAsStringInternal@QualType@clang@@SAXPEBVType@2@VQualifiers@2@AEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBUPrintingPolicy@2@@Z
    ?decls_begin@DeclContext@clang@@QEBA?AVdecl_iterator@12@XZ
    ?field_begin@RecordDecl@clang@@QEBA?AV?$specific_decl_iterator@VFieldDecl@clang@@@DeclContext@2@XZ
    ?errs@llvm@@YAAEAVraw_ostream@1@XZ
    ?grow_pod@SmallVectorBase@llvm@@IEAAXPEAX_K1@Z
    ?write@raw_ostream@llvm@@QEAAAEAV12@E@Z
    ?write@raw_ostream@llvm@@QEAAAEAV12@PEBD_K@Z
    ?castFromDeclContext@Decl@clang@@SAPEAV12@PEBVDeclContext@2@@Z
    ??1raw_ostream@llvm@@UEAA@XZ
    ??1raw_string_ostream@llvm@@UEAA@XZ
    ?flush_nonempty@raw_ostream@llvm@@AEAAXXZ
    ?getASTContext@Decl@clang@@QEBAAEAVASTContext@2@XZ
    ?handle@raw_ostream@llvm@@EEAAXXZ
    ?preferred_buffer_size@raw_ostream@llvm@@MEBA_KXZ
    ?write_impl@raw_string_ostream@llvm@@EEAAXPEBD_K@Z
    ?castToDeclContext@Decl@clang@@SAPEAVDeclContext@2@PEBV12@@Z
    ?classof@DeclContext@clang@@SA_NPEBVDecl@2@@Z
    ?setValueNoAlloc@internal@runtime@cling@@YAXPEAX00D_K@Z
	?setValueNoAlloc@internal@runtime@cling@@YAXPEAX00DM@Z
	cling_runtime_internal_throwIfInvalidPointer
  )

  #these symols are project specific (i.e this demo defines these symobls in cling-demo.cpp. They must be manually exported)
  set(cling_exports ${cling_exports}
  	?getAnotherGlobal@@YAMXZ
  	?setAnotherGlobal@@YAXM@Z
  	)

  if ($<CONFIG:Debug>)
    set(cling_exports ${cling_exports}
      ??$dyn_cast@VValueDecl@clang@@$$CBVDecl@2@@llvm@@YAPEBVValueDecl@clang@@PEBVDecl@2@@Z
    )
  endif()

  # List to '/EXPORT:sym0 /EXPORT:sym1 /EXPORT:sym2 ...'
  foreach(sym ${cling_exports})
    set(cling_link_str "${cling_link_str} /EXPORT:${sym}")
  endforeach(sym ${cling_exports})

  set_property(TARGET cling-demo APPEND_STRING PROPERTY LINK_FLAGS ${cling_link_str})

endif(MSVC)

# ...which links against clingInterpreter (and its dependencies).
target_link_libraries(cling-demo clingInterpreter)

# Provide LLVMDIR to cling-demp.cpp:
target_compile_options(cling-demo PUBLIC -DLLVMDIR="${LLVM_INSTALL_PREFIX}" -I${LLVM_INSTALL_PREFIX}/include)

set_target_properties(cling-demo
  PROPERTIES ENABLE_EXPORTS 1)
```

**Comment by bellenot** [comment]
Cool! Glad you managed to make it working! :slightly_smiling_face:

**Comment by danhambleton** [comment]
FYI, I sometimes still hit this error:

```
IncrementalExecutor::executeFunction: symbol '_emutls_v._Init_thread_epoch' unresolved while linking [cling interface function]!
You are probably missing the definition of _emutls_v._Init_thread_epoch
Maybe you need to load the corresponding shared library?
```

**Comment by bellenot** [comment]
Yes, I know, it's one of the errors I'm currently debugging (might take some time...)

**Comment by bellenot** [comment]
FYI, the `'_emutls_v._Init_thread_epoch'` error should be fixed now

**Comment by danhambleton** [comment]
Amazing, thanks!
